### ISSUE TITLE
Unfreezing all the layers for fine tuning

### ISSUE THREAD

**Issue by srm-soumya** [Microsoft/CNTK#2922]
I am trying to implement a transfer learning model for Image classification task. These are the steps I follow in Phase-1:

- Import a base layer and freeze it using CloneMethod.freeze
- Add 2 dense layers, followed by the output layer
- Train for some epochs
- Save the trained model

Once I have trained my dense layers for a few number of epochs, I want to train the entire network. So I tried doing the following steps in Phase-2:

- Load the trained model from Phase-1 using `CloneMethod.clone`

When I inspect the nodes of the trained-model, I see the `needs_gradient` property of the nodes that were freezed initially is still False. Is there a way to unfreeze it, so that I can train the entire network.

I had a look at this issue: https://github.com/Microsoft/CNTK/issues/2144

It is not working for me.
- Although I pass a name for the Input layer while cloning, it does not seem to retain it after the model is created, so there is no way to refer it. (Although I could do cntk.logging.find_all_with_name(model, '')[i])

One Suggestion: Will it make sense to add unique ID's to the layers if the user has not specified a name parameter for it. I have seen some pre-trained models in the CNTK repo (probably imported from caffe) where the names are missing for a lot of layers, it becomes little difficult to work with them.

**Comment by KeDengMS** [comment]
Another way is to create two learners, each with different set of parameters that you'd train.

**Comment by delzac** [comment]
Sorry, can i just confirm that there is no way to unfreeze a already frozen layer? I encountered the same situation.

**Comment by YanYas** [comment]
It seems that in some models, Parameters become Constants when saved. This isn't the case for all of them, but if you're refining your transfer learned-model, it could be that those parameters are now baked in.

One potential way around this is to get the [Constants](https://www.cntk.ai/pythondocs/cntk.ops.functions.html?highlight=constants#cntk.ops.functions.Function.constants) from the model, and for each Constant, create a list of Parameters which take the [Value ](https://www.cntk.ai/pythondocs/cntk.variables.html?highlight=value#cntk.variables.Constant.value) from the constant (an NDArrayView, which you pass into a new Value, in Dotnet).

Create a 'Replacements' dictionary in a loop where entries are added as each key being a Constant matched by each Parameter as a value. Then Clone(CloneMethod.Clone, Replacements). 

The output of the clone should be of parameters replacing all of those constants with the original values. Do this carefully of course. You may want to make sure that you are only effecting Constants that are weights or biases by checking their names first.

I wouldn't necessarily say that thi a way to make ONNX models trainable because they vary in origin which sometimes mean that they lack a batch axis or dynamic axis. And I haven't checked at that level yet ;-)
